src/error/error.h

```diff
@@ -0,0 +1,161 @@
+class Error {
+ public:
+  Error(const std::string& message) : message_(message) {}
+
+  virtual std::string toString() const { return message(); }
+
+ protected:
+  std::string message() const { return message_; }
+ private:
+  std::string message_;
+};
+
+// Either an error or a positive result.
+// This class is immutable.
+template <typename Err>
+class MaybeError {
+ public:
+  // Constructor for no error.
+  // Not explicit on purpose.
+  MaybeError() : is_ok_(true) {}
+
+  // Represent an error.
+  // Not explicit on purpose.
+  MaybeError(Err&& error) {
+    // Placement new.
+    new (&as_Err()) Err(std::move(error));
+  }
+
+  // Represent an error.
+  // Not explicit on purpose.
+  MaybeError(const Err& error) {
+    // Placement new.
+    new (&as_Err()) Err(error);
+  }
+
+  ~MaybeError() {
+    if (!ok())
+      as_Err().~Err();
+  }
+
+  // Check if it's an error or a value.
+  bool ok() const { return is_ok_; }
+
+  // Return the error if it is one, fail otherwise.
+  const Err& error_or_die() const {
+    if (ok())
+      throw std::domain_error("ErrorOr was value, asked for error");
+    return as_Err();
+  }
+
+  // Return the error if it is one, otherwise return "Ok".
+  std::string toString() const {
+    if (ok())
+      return "Ok";
+    else
+      return error_or_die().toString();
+  }
+
+ private:
+  Err& as_Err() {
+    return *reinterpret_cast<Err*>(&error_memory_);
+  }
+  const Err& as_Err() const {
+    return *reinterpret_cast<const Err*>(&error_memory_);
+  }
+  const bool is_ok_ = false;
+  std::array<char, sizeof(Err)> error_memory_;
+};
+
+// Either an error or a T.
+// This class is immutable (but the T can be mutable).
+template <typename T, typename Err = Error>
+class ErrorOr {
+ public:
+  // Not explicit on purpose.
+  ErrorOr(Err&& error) : is_ok_(false) {
+    // Placement new.
+    new (&as_error()) Err(error);
+  }
+
+  // Not explicit on purpose.
+  ErrorOr(const T& value) : is_ok_(true) {
+    // Placement new.
+    new (&as_T()) T(value);
+  }
+
+  // Not explicit on purpose.
+  ErrorOr(T&& value) : is_ok_(true) {
+    // Placement new.
+    new (&as_T()) T(std::move(value));
+  }
+
+  // Check if it's an error or a value.
+  bool ok() const { return is_ok_; }
+
+  // Return the value if it is one, fail otherwise.
+  T& value_or_die() {
+    if (!ok())
+      throw std::domain_error("ErrorOr was error, asked for value");
+    return as_T();
+  }
+
+  // Return the value if it is one, fail otherwise.
+  const T& value_or_die() const {
+    if (!ok())
+      throw std::domain_error("ErrorOr was error, asked for value");
+    return as_T();
+  }
+
+  // Return the error if it is one, fail otherwise.
+  const Err& error_or_die() const {
+    if (ok())
+      throw std::domain_error("ErrorOr was value, asked for error");
+    return as_error();
+  }
+
+  // Return the error if it is one, otherwise return "Ok".
+  std::string toString() const {
+    if (ok())
```

Comment:
I would add that you usually test the value case before the error case. So you should swap the two cases instead of changing the condition.